applications/asset_tracker_v2/src/cloud/cloud_codec/nrf_cloud_codec.c

```diff
@@ -0,0 +1,609 @@
+	if (err) {
+		LOG_ERR("Encoding error: %d returned at %s:%d", err, __FILE__, __LINE__);
+		goto exit;
+	}
+
+	err = add_meta_data(root_obj, APP_ID_BUTTON, &ui_buf->btn_ts);
+	if (err) {
+		LOG_ERR("Encoding error: %d returned at %s:%d", err, __FILE__, __LINE__);
+		goto exit;
+	}
+
+	buffer = cJSON_PrintUnformatted(root_obj);
+	if (buffer == NULL) {
+		LOG_ERR("Failed to allocate memory for JSON string");
+
+		err = -ENOMEM;
+		goto exit;
+	}
+
+	ui_buf->queued = false;
+
+	if (IS_ENABLED(CONFIG_CLOUD_CODEC_LOG_LEVEL_DBG)) {
+		json_print_obj("Encoded message:\n", root_obj);
+	}
+
+	output->buf = buffer;
+	output->len = strlen(buffer);
+
+	ui_buf->queued = false;
+
+exit:
+	cJSON_Delete(root_obj);
+	return err;
+}
+
+int cloud_codec_encode_batch_data(
+				struct cloud_codec_data *output,
+				struct cloud_data_gps *gps_buf,
+				struct cloud_data_sensors *sensor_buf,
+				struct cloud_data_modem_dynamic *modem_dyn_buf,
+				struct cloud_data_ui *ui_buf,
+				struct cloud_data_accelerometer *accel_buf,
+				struct cloud_data_battery *bat_buf,
+				size_t gps_buf_count,
+				size_t sensor_buf_count,
+				size_t modem_dyn_buf_count,
+				size_t ui_buf_count,
+				size_t accel_buf_count,
+				size_t bat_buf_count)
+{
+	int err;
+	char *buffer;
+	bool object_added = false;
+
+	cJSON *root_obj = cJSON_CreateObject();
+
+	if (root_obj == NULL) {
+		cJSON_Delete(root_obj);
+		return -ENOMEM;
+	}
+
+	err = json_common_batch_data_add(root_obj, JSON_COMMON_MODEM_DYNAMIC,
+					 modem_dyn_buf, modem_dyn_buf_count,
+					 DATA_MODEM_DYNAMIC);
+	if (err == 0) {
+		object_added = true;
+	} else if (err != -ENODATA) {
+		goto exit;
+	}
+
+	err = json_common_batch_data_add(root_obj, JSON_COMMON_GPS,
+					 gps_buf, gps_buf_count,
+					 DATA_GPS);
+	if (err == 0) {
+		object_added = true;
+	} else if (err != -ENODATA) {
+		goto exit;
+	}
+
+	err = json_common_batch_data_add(root_obj, JSON_COMMON_SENSOR,
+					 sensor_buf, sensor_buf_count,
+					 DATA_ENVIRONMENTALS);
+	if (err == 0) {
+		object_added = true;
+	} else if (err != -ENODATA) {
+		goto exit;
+	}
+
+	err = json_common_batch_data_add(root_obj, JSON_COMMON_UI,
+					 ui_buf, ui_buf_count,
+					 DATA_BUTTON);
+	if (err == 0) {
+		object_added = true;
+	} else if (err != -ENODATA) {
+		goto exit;
+	}
+
+	err = json_common_batch_data_add(root_obj, JSON_COMMON_BATTERY,
+					 bat_buf, bat_buf_count,
+					 DATA_BATTERY);
+	if (err == 0) {
+		object_added = true;
+	} else if (err != -ENODATA) {
+		goto exit;
+	}
+
+	err = json_common_batch_data_add(root_obj, JSON_COMMON_ACCELEROMETER,
+					 accel_buf, accel_buf_count,
+					 DATA_MOVEMENT);
+	if (err == 0) {
+		object_added = true;
+	} else if (err != -ENODATA) {
+		goto exit;
+	}
+
+	if (!object_added) {
+		err = -ENODATA;
+		LOG_DBG("No data to encode, JSON string empty...");
+		goto exit;
+	} else {
+		/* At this point err can be either 0 or -ENODATA. Explicitly set err to 0 if
+		 * objects has been added to the rootj object.
+		 */
+		err = 0;
```

Comment:
Should we `goto exit` here, as no data has been added?

Reply:
Actually, data may have been added in this case. If an earlier call to `json_common_batch_data_add` succeeds in adding a certain data type to root and a subsequent call returns `-ENODATA` for a given data type. (which was not queued). We want to prevent propagating an error code.

Reply:
This may be a little bit clunky. We should consider changing this in the future. I'm sure there is a cleaner way of doing it.

Reply:
Ah right, it's all coming back to me now.. Let's leave it as is for now